design/src/rfcs/rfc_0020_endpoints_20.md

```diff
@@ -0,0 +1,599 @@
+}
+```
+
+The trait itself would then be parameterized by service-specific endpoint parameter, eg:
+`aws_sdk_s3::endpoint::Params`. The endpoint parameters we would use for S3 (e.g. including `Bucket`) are different
+from the endpoint parameters we might use for a service like DynamoDB which, today, doesn't have any custom endpoint
+behavior.
+
+Going forward we want to provide two different avenues for customers to customize endpoints:
+
+1. Configuration driven URL override. This mechanism hasn't been specified, but suppose that the Rust SDK supported
+   an `SDK_ENDPOINT` environment variable. This variable would be an input to the existing endpoint resolver.
+   machinery and would be backwards compatible with other SDKs (e.g. by prefixing the bucket as a host label for S3).
+2. Wholesale endpoint resolver override. In this case, customers would gain access to all endpoint parameters and be
+   able to write their own resolver.
+
+This RFC proposes making the following changes:
+
+1. For the current _global_ ability to override an endpoint, instead of accepting an `AwsEndpoint`, accept a URI. This
+   will simplify the interface for most customers who don't actually need logic-driven endpoint construction. The
+   Endpoint that can be set will be passed in as the `SDK::Endpoint` built-in. This will be renamed to `endpoint_uri`
+   for clarity. **All** AWS services **MUST** accept the `SDK::Endpoint` built-in.
+2. For complex, service-specific behavior, customers will be able to provide a service specific endpoint resolver at
+   client construction time. This resolver will be parameterized with the service-specific parameters type, (
+   eg. `aws_sdk_s3::endpoint::Params`). Finally, customers will be able to access the `default_resolver()` for S3
+   directly. This will enable them to utilize the default S3 endpoint resolver in their resolver implementation.
+
+**Example: overriding the endpoint URI globally**
+
+```rust
+async fn main() {
+    let sdk_conf = aws_config::from_env().endpoint_uri("http://localhost:8123".parse().unwrap()).load().await;
+    let dynamo = aws_sdk_dynamodb::Client::new(&sdk_conf);
+    // snip ...
+}
+```
+
+**Example: overriding the endpoint resolver for a service**
+
+```rust
+/// Resolve to Localhost when an environment variable is set
+struct CustomDdbResolver;
+
+impl ResolveEndpoint<aws_sdk_dynamodb::endpoint::Params> for CustomDdbResolver {
+    fn resolve_endpoint(&self, params: &Params) -> Result<Endpoint, EndpointResolutionError> {
+        // custom resolver to redirect to DDB local if a flag is set
+        let base_endpoint = aws_sdk_dynamodb::endpoint::default_resolver().resolve_endpoint(params).expect("valid endpoint should be resolved");
+        if env::var("LOCAL") == Ok("true") {
+            // update the URI on the returned endpoint to localhost while preserving the other properties
+            base_endpoint.builder().uri("http://localhost:8888").build()
```

Comment:
What other properties?

Reply:
the `properties` which contain auth info e.g.